src/scheduler/worker.rs

```diff
@@ -140,22 +154,124 @@ impl<VM: VMBinding> GCWorker<VM> {
         work.do_work(self, self.mmtk);
     }
 
+    /// Entry of the worker thread.
+    /// Each worker will keep polling and executing work packets in a loop.
     pub fn run(&mut self, tls: VMWorkerThread, mmtk: &'static MMTK<VM>) {
         self.tls = tls;
         self.copy = crate::plan::create_gc_worker_context(tls, mmtk);
-        self.shared.parked.store(false, Ordering::SeqCst);
+        self.shared.parked.store(false, Ordering::Relaxed);
         loop {
-            while let Some((bucket, mut work)) = self.local_work_buffer.pop() {
-                debug_assert!(self.scheduler.work_buckets[bucket].is_activated());
-                work.do_work_with_stat(self, mmtk);
-            }
-            let mut work = self.scheduler().poll(self);
+            let mut work = self.poll();
             debug_assert!(!self.shared.is_parked());
             work.do_work_with_stat(self, mmtk);
         }
     }
 }
+
+/// A worker group to manage all the GC workers (except the coordinator worker).
+pub struct WorkerGroup<VM: VMBinding> {
+    /// Shared worker data
+    pub workers_shared: Vec<Arc<GCWorkerShared<VM>>>,
+    /// Handles for stealing packets from workers
+    pub stealers: Vec<Stealer<Box<dyn GCWork<VM>>>>,
+    parked_workers: AtomicUsize,
+}
+
+impl<VM: VMBinding> WorkerGroup<VM> {
+    /// Create a WorkerGroup
+    pub fn new(num_workers: usize) -> Arc<Self> {
+        let workers_shared = (0..num_workers)
+            .map(|_| Arc::new(GCWorkerShared::<VM>::new()))
+            .collect::<Vec<_>>();
+
+        let stealers = workers_shared
+            .iter()
+            .map(|worker| worker.local_work_buffer.stealer())
+            .collect();
+
+        Arc::new(Self {
+            workers_shared,
+            stealers,
+            parked_workers: Default::default(),
+        })
+    }
+
+    /// Spawn all the worker threads
+    pub fn spawn(
+        &self,
+        mmtk: &'static MMTK<VM>,
+        sender: Sender<CoordinatorMessage<VM>>,
+        tls: VMThread,
+    ) {
+        // Spawn each worker thread.
+        for (ordinal, shared) in self.workers_shared.iter().enumerate() {
+            let worker = Box::new(GCWorker::new(
+                mmtk,
+                ordinal,
+                mmtk.scheduler.clone(),
+                false,
+                sender.clone(),
+                shared.clone(),
+            ));
+            VM::VMCollection::spawn_gc_thread(tls, GCThreadContext::<VM>::Worker(worker));
+        }
+    }
+
+    /// Get the number of workers in the group
+    #[inline(always)]
+    pub fn worker_count(&self) -> usize {
+        self.workers_shared.len()
+    }
+
+    /// Increase the packed-workers counter.
+    /// Called before a worker is parked.
+    ///
+    /// Return true if all the workers are parked.
+    #[inline(always)]
+    pub fn inc_parked_workers(&self) -> bool {
+        let old = self.parked_workers.fetch_add(1, Ordering::Relaxed);
+        old + 1 == self.worker_count()
```

Comment:
I suggest adding assertions to the inc/dec methods to make sure `parked_workers` is always in the range `[0, worker_count())`.